Introduce new `VolumeAttachment` API Object

saad-ali opened the pull request:
**What this PR does / why we need it**:

Introduce a new `VolumeAttachment` API Object. This object will be used by the CSI volume plugin to enable external attachers (see design [here](https://github.com/kubernetes/community/pull/1258). In the future, existing volume plugins can be refactored to use this object as well.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:  Part of issue https://github.com/kubernetes/features/issues/178

**Special notes for your reviewer**:
None

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

jsafrane commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -121,3 +121,51 @@ func validateAllowVolumeExpansion(allowExpand *bool, fldPath *field.Path) field.
+		allErrs = append(allErrs, field.Invalid(fldPath, nodeName, msg))
+	}
+	return allErrs
+}
```
There should be also `ValidateVolumeAttachmentUpdate` that makes sure that Spec can't be changed.

jsafrane replied:
This is fixed now few lines below.

jsafrane reviewed (commented):
I miss the object registration in pkg/registry/storage/volumeattachment.. IMO it should be in this PR - the API object is unusable without it.

jsafrane commented:
@kubernetes/api-reviewers @kubernetes/sig-storage-api-reviews

liggitt commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
+	// Indicates the volume is successfully attached.
+	// This field must only be set by the entity completing the attach
+	// operation, i.e. the external-attacher.
+	IsAttached bool
```
a couple things to consider from the [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md):
> * Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options. Plan for future expansions by describing the policy options explicitly as a string type alias (e.g. TerminationMessagePolicy).
> * The name of a field expressing a boolean property called 'fooable' should be called Fooable, not IsFooable.

liggitt replied:
Also, it seems like we'd need to be able to tell from status whether attach had begun.

Since attach is not instant, there could be a period of time during 

1. Kubernetes creates VolumeAttachment
2. Attacher observes VolumeAttachment, starts attaching
3. PV/VolumeAttachment objects are vulnerable to deletion/finalization, since status doesn't indicate the volume is attached or is attaching
4. Attacher completes attaching, tries to update status to indicate attached state, object is gone

It seems like it should be possible to tell where we are in the `detached -> attaching -> attached -> detaching -> detached` lifecycle by looking at status (and making the attacher do a successful status update from `detached -> attaching` before starting the attach operation prevents races trying to attach a volume that is getting deleted)

jsafrane replied:
The idea is that we do not need "attaching" state, simple presence of the object with `(Is)Attached==false` is enough. The same for "detaching", presence of `deletetionTimestamp` is enough.

> `detached` -> `attaching`
This is not possible. VolumeAttachment with `deletetionTimestamp` must be deleted first in API server and Kubernetes must create a new instance. Or is it allowed to cancel deletetionTimestamp? It looks odd.

jsafrane replied:
We did not want to introduce state machines with detached -> attaching -> attached -> detaching -> detached because it's not possible to add a new state there later.

liggitt replied:
>> detached -> attaching
>
> This is not possible

Maybe I used the wrong term. Is `unattached -> attaching -> attached -> detaching -> detached` more correct? Don't we need to represent the state where the attacher has observed the request for the attachment and is currently working to attach (so the VolumeAttachment object should not be removed), but it has not yet attached (so the volume is not ready to mount)?

jsafrane replied:
I don't think we need to capture the state that attacher has started attaching. A/D controller with time out eventually and send some event to pod. There is a glitch though, there is no way how to tell if external attacher does not run or it's just slow.

liggitt commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
+	Attacher string
+
+	// The volume to attach.
+	Volume api.VolumeSource
```
Is it intentional that VolumeAttachment is a cluster-scoped type that references `api.VolumeSource` instead of `api.PersistentVolumeSource`?

jsafrane replied:
Yes. It serves as memento of pod's VolumeSource in case the pod is deleted and we need to detach its volumes.

liggitt replied:
there's no place to indicate the original namespace, pod name or pod uid... is none of that information needed to resolve the volume or referenced secrets?

jsafrane replied:
We moved to supporting PersistentVolumes only and we use namespaced SecretReferences there

gnufied commented:
cc @kubernetes/sig-storage-api-reviews

jsafrane commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
+	// This field must only be set by the entity completing the attach
+	// operation, i.e. the external-attacher.
+	// +optional
+	AttachError VolumeError
```
Should it be a pointer? `kubectl get volumeattachment` shows ugly output when I created an instance with no Status:

```
  apiVersion: storage.k8s.io/v1
  kind: VolumeAttachment
  metadata:
    creationTimestamp: 2017-10-24T16:26:47Z
    name: first
    namespace: ""
    resourceVersion: "341"
    selfLink: /apis/storage.k8s.io/v1/volumeattachments/first
    uid: 2192999e-b8d8-11e7-b26b-525400224e72
  spec:
    attacher: csi/example
    nodeName: localhost
    volume:
      hostPath:
        path: /tmp
        type: ""
  status:
    attachError:
      Message: ""
      time: null
    detachError:
      Message: ""
      time: null
    isAttached: false
kind: List
```

liggitt replied:
optional structs should be pointers - https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required

rootfs commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
+	// This field must only be set by the entity completing the attach
+	// operation, i.e. the external-attacher.
+	// +optional
+	AttachError VolumeError
```
Prefer a status slice similar to Pod `Conditions`. It helps track the evolution attachment status.

jsafrane replied:
We discussed that and there are rumors that Conditions are bad design and we should not use them for new features. I am not sure who's source of that rumor, @thockin?

saad-ali replied:
This was a result of discussions with @thockin and @erictune

kokhang commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
+	// This field must only be set by the entity completing the detach
+	// operation, i.e. the external-attacher.
+	// +optional
+	DetachError VolumeError
```
Is the VolumeAttachment object used as more like record/status keeping object or used to convey a desired state?
If the latter, then im assuming the detach is triggered when this VolumeAttachment object is deleted. So if an failure happens during the detach, the external-attacher will not have this object to specify the DetachError anymore.

jsafrane replied:
* VolumeAttachment is used for both - `Spec` for the desired state and `Status` for record keeping.
* We use finalizers to keep "deleted" VolumeAttachment instances around until associated volume is truly detached from a node. See https://github.com/kubernetes/community/pull/1258 for details.

liggitt replied:
are VolumeAttachment objects 1:1 with PV objects?

jsafrane replied:
No, one volume can be attached to multiple nodes and thus will have multiple VolumeAttachment instances.

liggitt replied:
so they are unique by (pv,node)? is that enforced by convention or by pinning the name format to something like `<pvName>.<nodeName>`, etc?

jsafrane replied:
Yes, it's just an convention. IMO, nobody but Kubernetes A/D controller should be allowed to create this object.

kokhang commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
+	Volume api.VolumeSource
+
+	// The node that the volume should be attached to.
+	NodeName string
```
How is the VolumeAttachment object used during a multi-attachments scenario? For example with RO volumes, will there be multiple of these VolumeAttachment objects for every RO "attachment" even if the volume source and node is the same or will there just be one?

kokhang replied:
Should we also store information about pods and mountDir in the case where plugins are not using the attacher-detacher controller?

jsafrane replied:
> volumes, will there be multiple of these VolumeAttachment objects for every RO "attachment" even if the volume source and node is the same or will there just be one?

We do not plan to change current behavior, i.e. a volume is attached to a node only once. So there will be one instance per node-volume pair. Kubelet then mounts the single device to individual pods (using bind-mounts).

> Should you also store information about pods and mountDir in the case where plugins are not using the attacher-detacher controller?

I am not sure I understand the question. That's kubelet job and is not related to volume attachment.

kokhang replied:
I was alluding to the fact that in non-attachable plugins, kubelet mount() does everything (attach, mountDevice and mount podDir). So my question was if there is a need to store pod information to determine when it is safe to detach (delete VolumeAttach object).

Look like kubelet can determine how many bind mount ref are there during a detach and delete the attachment once there are no bind mount left.

jsafrane replied:
`VolumeAttachment` is used only for attachable plugins (CSI). See the [proposal](https://github.com/kubernetes/community/pull/1258) for handling of CSI plugins that don't support `ControllerPublishVolume` call.

And indeed, kubelet has its own means how to track pods that use the volume and it won't let Attach/Detach controller to detach a volume that's still mounted.

jsafrane commented:
I updated the API to current version in the proposal + added its registration and validation. `kubect get va`  and `kubectl create -f volumeattachment.yaml` now works.

thockin commented:
We should be judicious with shortnames like `va` - considering this is not something humans will use very often, I'd propose to not add a shortcut

thockin commented on pkg/apis/storage/types.go:
```diff
@@ -80,3 +80,110 @@ type StorageClassList struct {
+// Right now only PersistenVolumes can be attached via external attacher,
+// in future we may allow also inline volumes in pods.
+// Exactly one member can be set.
+type AttachedVolumeSource struct {
```
In general, we've swung towards discriminated unions.  Given that this has just one item, I suppose we can forego the discriminator for now, but as soon as we add a second option, we need it.  We can add a `type` field that defaults to "PersistentVolume" when we need it.

saad-ali replied:
Ack.

thockin commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -121,3 +127,99 @@ func validateAllowVolumeExpansion(allowExpand *bool, fldPath *field.Path) field.
+
+	// Spec is read-only
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "spec is immutable after creation"))
```
should just say "field is immutable"

saad-ali replied:
Changing

thockin reviewed (commented):
From other doc:

Did we rule out simply storing this as fields on PV?

saad-ali commented:
@caesarxuchao `./hack/test-update-storage-objects.sh` is failing with `panic: storage codec doesn't seem to match given type: Internal type not encodable: no kind "VolumeAttachment" is registered for version "storage.k8s.io/v1beta1"`. Do we have to add this new object to both `storage.k8s.io/v1beta1` and `storage.k8s.io/v1`? Or can we get away with just adding it to `storage.k8s.io/v1`?

saad-ali commented:
> We should be judicious with shortnames like va - considering this is not something humans will use very often, I'd propose to not add a shortcut

Ack. Removing.

> From other doc: Did we rule out simply storing this as fields on PV?

Mostly because it makes detaching tricker.

saad-ali committed:
Introduce storage v1alpha1 and VolumeAttachment

Introduce the v1alpha1 version to the Kubernetes storage API. And add a
new VolumeAttachment object to that version. This object will initially
be used only by the new CSI Volume Plugin. Eventually existing volume
plugins can be refactored to use it too.

saad-ali committed:
Generated files for new VolumeAttachemnt object

jsafrane commented on staging/src/k8s.io/api/storage/v1alpha1/types.go:
```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
wrong year

saad-ali replied:
Fixed.

jsafrane commented on staging/src/k8s.io/api/storage/v1alpha1/types.go:
```diff
@@ -0,0 +1,126 @@
+// to/from the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
```
I am not sure what's the process here, should we add also StorageClass to v1alpha1 api? It looks like we're removing it.

saad-ali replied:
Nope according to @lavalamp we just need to introduce the new VolumeAttachment object in v1alpha1 we can leave the StorageClass objects in v1beta1 and v1.

CaoShuFeng replied:
Seems has some relationship with https://github.com/kubernetes/kubernetes/pull/54463#issuecomment-345605558

jsafrane reviewed (commented):
lgtm to VolumeType API, however I am not sure if v1alpha1 is the right version. We dropped that API a while ago, should we use v2alpha1 or v1alpha2?

jsafrane commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -121,3 +127,99 @@ func validateAllowVolumeExpansion(allowExpand *bool, fldPath *field.Path) field.
+		allErrs = append(allErrs, field.Required(fldPath, attacher))
+	}
+	if len(attacher) > 0 {
+		for _, msg := range validation.IsQualifiedName(strings.ToLower(attacher)) {
```
Why qualified name? CSI does not describe any format.

saad-ali replied:
You're right. I removed it for now. Once https://github.com/container-storage-interface/spec/issues/145 closes, we can reintroduce it.

jsafrane reviewed (commented):
lgtm in general, with one minor nit.

It needs review from someone who understands API registration though.

saad-ali commented:
> It needs review from someone who understands API registration though.

@caesarxuchao was a huge help with this PR! He helped me figure out how to make etcd storage for the new v1alpha1 storage group work. Adding him for API machinery LGTM.

@thockin for API review (this mirrors what we agreed on in the the [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md).

saad-ali commented:
Cross build failure is due to unrelated issue being tracked here: https://github.com/kubernetes/kubernetes/issues/55711

thockin commented:
This is a ludicrous amount of work to add a new API.  Absolutely insane.
/lgtm
/approve

caesarxuchao commented on pkg/registry/storage/volumeattachment/strategy.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func (volumeAttachmentStrategy) AllowUnconditionalUpdate() bool {
+	return true
```
Do you really want this to be true?

It will defeat the optimistic concurrency built around resourceVersion. See https://github.com/kubernetes/kubernetes/blob/02b3928a921b2e0930bf67b48f71c9dcd1086b4f/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L574-L579

saad-ali replied:
Nope. Changed it to false.

caesarxuchao commented on pkg/registry/storage/volumeattachment/strategy.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func (volumeAttachmentStrategy) ValidateUpdate(ctx genericapirequest.Context, obj, old runtime.Object) field.ErrorList {
+	errorList := validation.ValidateVolumeAttachment(obj.(*storage.VolumeAttachment))
```
nit: add a variable and assign `obj.(*storage.VolumeAttachment)` to it, to save another type assertion.

saad-ali replied:
Done.

caesarxuchao commented on staging/src/k8s.io/api/storage/v1alpha1/register.go:
```diff
@@ -0,0 +1,53 @@
+	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
+	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
+	SchemeBuilder      = runtime.NewSchemeBuilder(addKnownTypes)
+	localSchemeBuilder = &SchemeBuilder
```
I know you copied the pattern from some other files, but still please remove the `localSchemeBuilder`, it's unnecessary. Also please remove the TODO at line 37.

saad-ali replied:
Will do.

caesarxuchao reviewed (commented):
A few comments and then the apimachinery related code lgtm.

saad-ali commented:
@caesarxuchao feedback addressed, PTAL.

caesarxuchao commented:
apimachinery part lgtm.

jsafrane commented:
With hack/local-up-cluster.sh, I don't see storage.k8s.io/v1alpha1 API:

```
$ curl 127.0.0.1:8080/apis/storage.k8s.io/v1alpha1
{                                                                              
  "kind": "Status",                                                            
  "apiVersion": "v1",                                                          
  "metadata": {                                                                
                                                                               
  },                                                                           
  "status": "Failure",                                                         
  "message": "the server could not find the requested resource",               
  "reason": "NotFound",                                                        
  "details": {                                                                 
                                                                               
  },                                                                           
  "code": 404                                                                  
}
```

Do I need to enable it somewhere?

liggitt commented:
alpha API groups are not enabled by default. you can enable it with `--runtime-config=apis/storage.k8s.io/v1alpha1=true` (`RUNTIME_CONFIG=apis/storage.k8s.io/v1alpha1=true` envvar in local up cluster, I think)

CaoShuFeng commented:
Hi, I don't know why CI didn't find this, But I get this error:
```
[fujitsu@localhost kubernetes]$ make test WHAT=k8s.io/kubernetes/pkg/registry/storage/storageclass/storage
Running tests for APIVersion: v1,admissionregistration.k8s.io/v1alpha1,admission.k8s.io/v1alpha1,apps/v1beta1,apps/v1beta2,apps/v1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2beta1,batch/v1,batch/v1beta1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,policy/v1beta1,rbac.authorization.k8s.io/v1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,scheduling.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,
+++ [1120 14:38:40] Running tests without code coverage
2017-11-20 14:38:43.201383 I | integration: launching 7874548005808368453 (unix://localhost:78745480058083684530)
2017-11-20 14:38:43.201637 I | etcdserver: name = 7874548005808368453
2017-11-20 14:38:43.201653 I | etcdserver: data dir = /tmp/etcd404175323
2017-11-20 14:38:43.201657 I | etcdserver: member dir = /tmp/etcd404175323/member
2017-11-20 14:38:43.201660 I | etcdserver: heartbeat = 10ms
2017-11-20 14:38:43.201663 I | etcdserver: election = 100ms
2017-11-20 14:38:43.201665 I | etcdserver: snapshot count = 0
2017-11-20 14:38:43.201671 I | etcdserver: advertise client URLs = unix://127.0.0.1:2100224704
2017-11-20 14:38:43.201676 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:2100124704
2017-11-20 14:38:43.201692 I | etcdserver: initial cluster = 7874548005808368453=unix://127.0.0.1:2100124704
2017-11-20 14:38:43.208654 I | etcdserver: starting member 7f7f4f96ec3c19c6 in cluster 787606054cdf7e02
2017-11-20 14:38:43.208687 I | raft: 7f7f4f96ec3c19c6 became follower at term 0
2017-11-20 14:38:43.208708 I | raft: newRaft 7f7f4f96ec3c19c6 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2017-11-20 14:38:43.208720 I | raft: 7f7f4f96ec3c19c6 became follower at term 1
2017-11-20 14:38:43.209874 I | etcdserver: set snapshot count to default 10000
2017-11-20 14:38:43.209918 I | etcdserver: starting server... [version: 3.1.10, cluster version: to_be_decided]
2017-11-20 14:38:43.210437 I | etcdserver/membership: added member 7f7f4f96ec3c19c6 [unix://127.0.0.1:2100124704] to cluster 787606054cdf7e02
2017-11-20 14:38:43.210826 I | integration: launched 7874548005808368453 (unix://localhost:78745480058083684530)
2017-11-20 14:38:43.238973 I | raft: 7f7f4f96ec3c19c6 is starting a new election at term 1
2017-11-20 14:38:43.238991 I | raft: 7f7f4f96ec3c19c6 became candidate at term 2
2017-11-20 14:38:43.239018 I | raft: 7f7f4f96ec3c19c6 received MsgVoteResp from 7f7f4f96ec3c19c6 at term 2
2017-11-20 14:38:43.239034 I | raft: 7f7f4f96ec3c19c6 became leader at term 2
2017-11-20 14:38:43.239041 I | raft: raft.node: 7f7f4f96ec3c19c6 elected leader 7f7f4f96ec3c19c6 at term 2
2017-11-20 14:38:43.239199 I | etcdserver: published {Name:7874548005808368453 ClientURLs:[unix://127.0.0.1:2100224704]} to cluster 787606054cdf7e02
2017-11-20 14:38:43.239262 I | etcdserver: setting up the initial cluster version to 3.1
2017-11-20 14:38:43.239352 N | etcdserver/membership: set the initial cluster version to 3.1
2017-11-20 14:38:43.239384 I | etcdserver/api: enabled capabilities for version 3.1
2017-11-20 14:38:43.243130 I | integration: terminating 7874548005808368453 (unix://localhost:78745480058083684530)
2017-11-20 14:38:43.255072 I | integration: terminated 7874548005808368453 (unix://localhost:78745480058083684530)
--- FAIL: TestCreate (0.05s)
	resttest.go:320: Unexpected error: no kind "StorageClass" is registered for version "storage.k8s.io/v1alpha1"
2017-11-20 14:38:43.255365 I | integration: launching 5776451847361156600 (unix://localhost:57764518473611566000)
2017-11-20 14:38:43.255607 I | etcdserver: name = 5776451847361156600
2017-11-20 14:38:43.255617 I | etcdserver: data dir = /tmp/etcd561142654
2017-11-20 14:38:43.255623 I | etcdserver: member dir = /tmp/etcd561142654/member
2017-11-20 14:38:43.255627 I | etcdserver: heartbeat = 10ms
2017-11-20 14:38:43.255631 I | etcdserver: election = 100ms
2017-11-20 14:38:43.255635 I | etcdserver: snapshot count = 0
2017-11-20 14:38:43.255641 I | etcdserver: advertise client URLs = unix://127.0.0.1:2100424704
2017-11-20 14:38:43.255646 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:2100324704
2017-11-20 14:38:43.255654 I | etcdserver: initial cluster = 5776451847361156600=unix://127.0.0.1:2100324704
2017-11-20 14:38:43.263060 I | etcdserver: starting member 53e079aca545846c in cluster d2a46b8e8af27f5e
2017-11-20 14:38:43.263082 I | raft: 53e079aca545846c became follower at term 0
2017-11-20 14:38:43.263089 I | raft: newRaft 53e079aca545846c [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2017-11-20 14:38:43.263093 I | raft: 53e079aca545846c became follower at term 1
2017-11-20 14:38:43.263868 I | etcdserver: set snapshot count to default 10000
2017-11-20 14:38:43.263890 I | etcdserver: starting server... [version: 3.1.10, cluster version: to_be_decided]
2017-11-20 14:38:43.264308 I | integration: launched 5776451847361156600 (unix://localhost:57764518473611566000)
2017-11-20 14:38:43.264413 I | etcdserver/membership: added member 53e079aca545846c [unix://127.0.0.1:2100324704] to cluster d2a46b8e8af27f5e
2017-11-20 14:38:43.353276 I | raft: 53e079aca545846c is starting a new election at term 1
2017-11-20 14:38:43.353322 I | raft: 53e079aca545846c became candidate at term 2
2017-11-20 14:38:43.353334 I | raft: 53e079aca545846c received MsgVoteResp from 53e079aca545846c at term 2
2017-11-20 14:38:43.353350 I | raft: 53e079aca545846c became leader at term 2
2017-11-20 14:38:43.353366 I | raft: raft.node: 53e079aca545846c elected leader 53e079aca545846c at term 2
2017-11-20 14:38:43.353547 I | etcdserver: published {Name:5776451847361156600 ClientURLs:[unix://127.0.0.1:2100424704]} to cluster d2a46b8e8af27f5e
2017-11-20 14:38:43.353623 I | etcdserver: setting up the initial cluster version to 3.1
2017-11-20 14:38:43.353852 N | etcdserver/membership: set the initial cluster version to 3.1
2017-11-20 14:38:43.360294 I | integration: terminating 5776451847361156600 (unix://localhost:57764518473611566000)
2017-11-20 14:38:43.379161 I | integration: terminated 5776451847361156600 (unix://localhost:57764518473611566000)
--- FAIL: TestUpdate (0.12s)
	resttest.go:458: unexpected error: no kind "StorageClass" is registered for version "storage.k8s.io/v1alpha1"
	resttest.go:463: unexpected error: storageclasses.storage.k8s.io "foo2" not found
panic: interface conversion: runtime.Object is nil, not *storage.StorageClass [recovered]
	panic: interface conversion: runtime.Object is nil, not *storage.StorageClass

goroutine 200 [running]:
testing.tRunner.func1(0xc42036e960)
	/usr/local/go/src/testing/testing.go:711 +0x2d2
panic(0x154b6e0, 0xc420324600)
	/usr/local/go/src/runtime/panic.go:491 +0x283
k8s.io/kubernetes/pkg/registry/storage/storageclass/storage.TestUpdate.func1(0x0, 0x0, 0x14, 0xc42116dda8)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/storage/storageclass/storage/storage_test.go:95 +0x147
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest.(*Tester).testUpdateEquals(0xc4202ba570, 0x21637a0, 0xc4202726e0, 0xc420613ed8, 0xc4201d45a0, 0x17c2d40)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go:465 +0x2bd
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest.(*Tester).TestUpdate(0xc4202ba570, 0x21637a0, 0xc420272580, 0xc42116ded8, 0xc4201d45a0, 0x17c2d40, 0xc420528140, 0x1, 0x1)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go:164 +0x61
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/testing.(*Tester).TestUpdate(0xc4201d4580, 0x21637a0, 0xc420272580, 0x17c2d40, 0xc420613f88, 0x1, 0x1)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/testing/tester.go:92 +0x209
k8s.io/kubernetes/pkg/registry/storage/storageclass/storage.TestUpdate(0xc42036e960)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/storage/storageclass/storage/storage_test.go:90 +0x362
testing.tRunner(0xc42036e960, 0x17c2d50)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de
FAIL	k8s.io/kubernetes/pkg/registry/storage/storageclass/storage	0.236s
Makefile:182: recipe for target 'test' failed
make: *** [test] Error 1
```

CaoShuFeng commented:
ref:
https://github.com/kubernetes/kubernetes/pull/56136
